### ISSUE TITLE
Does not type check on predicate of `if` statement when mapping a Set within the predicate

### ISSUE THREAD

**Issue by gianlo** [lampepfl/dotty#17597]
## Compiler version

3.2.2

## Minimized code
When mapping a set inline within the predicate of an if statement the type check does not happen.

The following code compiles:
```
val someInts = Set(1, 2, 3)
val question = if (someInts.map( 2 * _ ).contains("a")) "YES" else "NO"
println(question)
```

## Output
When that code runs it prints out "NO" (but I suspect this might be a random outcome).

## Expectation
The same code does not compile on scala 2.13.10.

If the mapping happens outside the predicate of the 'if' statement than it does not compile as expected.

Basically the code should not compile whether the mapping is done within or outside the 'if' statement predicate.

BTW this has lead to a subtle bug in our code that manifested in flaky tests (hence the suspicion that the result might be random).

**Comment by wjoel** [comment]
Minimization (I think), explicitly using `Set[Int]` results in the expected error message:
```scala
scala> Set(1).contains("a")
val res0: Boolean = false
                                                                                                                                                                                               
scala> Set[Int](1).contains("a")
-- [E007] Type Mismatch Error: -------------------------------------------------
1 |Set[Int](1).contains("a")
  |                     ^^^
  |                     Found:    ("a" : String)
  |                     Required: Int
  |
  | longer explanation available when compiling with `-explain`
1 error found
```

Scala 2.13.6 gives the expected type mismatch:
```scala
scala> Set(1).contains("a")
                       ^
       error: type mismatch;
        found   : String("a")
        required: Int
```

**Comment by gianlo** [comment]
I find puzzling that the following code
`
val someInts = Set(1, 2, 3)
val question = if (someInts.contains("a")) "YES" else "NO"
println(question)
`
does NOT compile in scala 3.2.2 and does the correct type check...
only when mapping inside the predicate clause it does not perform the type check :-(

**Comment by som-snytt** [comment]
Dotty needs `-Winfer-any` and it must include `-Winfer-matchable`.

In Scala 2, `-Xlint:infer-any`.

**Comment by odersky** [comment]
If you compile with -Xprint:typer, you can see what gets inferred:
```scala
    val question: String =
      if 
        someInts.map[Int | String](
          {
            def $anonfun(_$1: Int): Int | String = 2.*(_$1)
            closure($anonfun)
          }
        ).contains("a")
       then "YES" else "NO"
```
That explains the behavior. Scala 2 infers type variables with less context, which means an error is issued. 

This also shows that warning when widening to Any would not make any difference here.

**Comment by odersky** [comment]
Multiversal equality is designed to avoid problems like that. If `Set` was defined like this:

```scala
class Set[A] derives CanEqual:
  def contains[B](x: B)(using CanEqual[A, B]) = ???
```
Then you would get the proper error:
```
-- [E172] Type Error: test.scala:9:28 ------------------------------------------
9 |  s.map(2 * _).contains("a")
  |                            ^
  |           Values of types Int and String cannot be compared with == or !=
```
Unfortunately, to define `Set` like this we need to upgrade the Scala 2.13 standard library, and that has not happened yet.
There might be a way to retro-fit Scala 3 features like this to the Scala 2 library in a forwards and backwards compatible way while keeping the nice property that the library can be used from both Scala 2 and Scala 3. But it will be tricky.

**Comment by gianlo** [comment]
@odersky just to play that back, the issue I found is the expected behavior of the dotty compiler, i.e. it has inferred `Int | String` when mapping the Set of Int since the map was followed by the contain operator called with a String.

And your previous comment, suggests that since the compiler has inferred `Int | String`, this scenario would not be caught as a compiler warning (since it's not an Any).

If that's the case, from a user perspective, this behavior can unfortunately introduce bugs that are hard to find.


PS I 2x checked and the following
```
Set(1, 2, 3).map(identity).contains("a")
```
does compile in scala 3.2.2 outside the if statement.

**Comment by odersky** [comment]
Yes, all correct. It's an unfortunate side effect of more flexible type inference. The way to fix it is to get multi-versal equality into the standard library but this will be hard and will take time.

**Comment by som-snytt** [comment]
The proverbial 3rd party linter could identify well-known API that are not safe with respect to `CanEqual` and more precisely detect dubious inferred type args.

**Comment by gianlo** [comment]
thanks to all.
I'll inform my team about this so that we can be more vigilant :-)

**Comment by odersky** [comment]
As a stopgap measure we might want to extend the linter with warnings for cases like these. Or implement magic in the compiler to make it do multiversal equality checks for certain stdlib methods even if none are declared. I agree it's an annoying problem.

**Comment by odersky** [comment]
> Maybe it's a bit deeper issue than a lack of multiversal equality support from the collection library :-(

Maybe there is, but it is not related to the compiler or type inference. Type inference is correct here. It worked in Scala 2 only because of a _shortcoming_ in that not enough context was considered for picking type arguments here.

I believe the deeper issue if that for something like `contains` we do not want plain subtype polymorphism because of widenings.  A better design for `contains` is via type classes, and multi-versal equality implements that. 

The inference is a perfectly good one, and

**Comment by gianlo** [comment]
I think as a practical solution for us, is to declare the type of the Set when binding it to a variable. 
That way the widening does not be kick in and the 'misuse' of the `contains` method is caught at compilation time.

(I 2x checked this and it actually fails to compile if I declare the Set[Int] at variable binding).

Thanks again for the prompt replies :-)

**Comment by odersky** [comment]
Yes, explicit types work. In fact you can also just define the set as a different val before calling contains:

```scala
val s = Set(1, 2, 3)
val s2 = s.map(_ + 1)
s2.contains("a") // error
```
So the rule could be: Never chain `contains` in a sequence of operations.

**Comment by gianlo** [comment]
a-ha so maybe that's the one for the linter :-)

Are there other `collection` methods that might be affected by this pattern? 
It might be useful to add all of those to the linter then :-)

**Comment by som-snytt** [comment]
Under `-Winfer-union`:

```
3 |val question = if (someInts.map( 2 * _ ).contains("a")) "YES" else "NO"
  |                   ^^^^^^^^^^^^
  |                A type argument was inferred to be union type Int | String
  |                This may indicate a programming error.
1 warning found
```
